### ISSUE TITLE
Wayland frontend and server hang rendering threads when the screen turns off

### ISSUE THREAD

**Issue by UniversalSuperBox** [canonical/mir#2136]
The TMI version of this issue's investigation for Lomiri is https://gitlab.com/ubports/core/lomiri-system-compositor/-/issues/11

TL;DR: When the display turns off on a Mesa system, the mir_frontend's `swap_buffers` will hang after one successful frame. This is because Mesa throttles Wayland frame callbacks to no more than one per vsync, and the vsync has stopped when the display is off. Even if `eglSwapInterval` is set to 0, frame callbacks are throttled.

This seems to be expected behavior from a Wayland compositor, and in fact Mir is much more lenient than other compositors: opinionated compositors like wlroots and Clutter (not so sure about Clutter, but it seems to be part of the folklore online) will stop frame callbacks whenever a client is completely occluded by another client. This is efficient I suppose, saving CPU and GPU resources in buffer swaps and continuous rendering for a client that the user can't see. So I believe that we should not try and change this behavior with secret background timeouts in `swap_buffers` or the like.

However, every interesting Wayland application and toolkit has had to work around this problem since no other compositing protocol on the planet seems to do this. [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1489902) and [Qt](https://bugreports.qt.io/browse/QTBUG-69077) are two examples. [Qt's workaround](https://github.com/qt/qtwayland/commit/85bb158ddf08aca4d76c13c6a9fcd2637d84d3ea) involves timing out the wait for the frame callback and cancelling any further rendering, but I don't think that's a viable solution here. While it's not okay for your web browser or Qt application to hang up because the render thread is blocked on eglSwapBuffers, that's a design decision in many applications. The function which contains eglSwapBuffers often contains other side effects. Without eglSwapBuffers and those side effects completing as assumed, it might not be safe to advance the rendering or main threads.

This issue presents in `egmde` running under Sway since wlroots stops sending frame callbacks to occluded applications. To reproduce the problem, run `egmde` in a terminal (you can open your terminal with Windows+Enter). Move the `egmde` window to another workspace (Windows+Shift+2). Send an interrupt to the egmde process with Control+C in the terminal. Notice the egmde will never actually shut down.

There is some upstream work in Wayland to resolve this problem. [surface-suspension](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/99) is a protocol which allows the compositor to tell the client when its surface is suspended and it will no longer receive any frame callbacks. In the world as intended by this proposal, it seems that we would want all clients on a given (screen? display? display_server?) to be given the suspended signal when vsync is expected to stop.

**Comment by UniversalSuperBox** [comment]
Some more background:

Some Wayland developers blame the EGL protocol for this behavior, and would rather implement a nonstandard extension for eglSwapBuffers with a timeout in Mesa: https://gitlab.freedesktop.org/mesa/mesa/-/issues/4932. This is not actually a full solution, since Qt will just run slowly in that case.

The surface-suspension protocol is hotly debated because the Wayland developers can't decide whether breaking every EGL application ever is a good idea or not.

Once surface-suspension is implemented, it *might* be handled directly in Mesa's EGL implementation instead of needing to be implemented by the clients. But that's hotly debated too.

**Comment by peat-psuwit** [comment]
As I have read some Mir code, I think I have some information. But first, a small correction:

> Even if eglSwapInterval is set to 0, frame callbacks are throttled.

He might confused frame callback with wl_display's "sync", which should return pretty much immediately if the compositor is running. Mesa will wait for the later if the swap interval is 0 and should not block. See [here](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/egl/drivers/dri2/platform_wayland.c#L1132) and [here](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/99#note_967216).

---

If I understand the code correctly, throttling to v-sync in Mir is (ultimately) done in a platform[1]-specific way by blocking in either `RenderTarget::swap_buffers()` or `DisplaySyncGroup::post()`. For Wayland platform, this is done by waiting for Wayland's frame callback in `swap_buffers()`.

With the current implementation, Mir's rendering thread will wait indefinitely for the frame callback, with no timeout and no way to interrupt it. Adding timeout like in the Qt's Wayland QPA could be a good first step, but is not enough to fix the whole problem because:

- We will still waste CPU and GPU power rendering frames.
- As we keep doing that, Mir or Mir's user (e.g. QtMir's Qt main event loop) will be able to process the events once in every 100ms or so. Not the end of the world, but not very preferable.

The surface suspension Wayland extension could be useful here. We can wait for either of them to arrive to stop waiting. Then, if the frame callback times out (which we can detect now) or the surface is suspended, tell Mir that we're not shown and can stop compositing.

The bigger question then becomes: how do a platform tells Mir to stop compositing?

An obvious solution is probably the `pause()` and `resume()` callbacks passed from the `DisplayServer` class. This will stop various things including compositing. However, this will stop compositing for the whole server. Notice that frame callback is a property of a `wl_surface`, which means that if you have 2 windows and e.g. one of them is minimized, potentially only the minimized one will have frame callback suspended. Preferably, we should stop compositing for only a single display/output. I'm not sure if the scenario can actually happen, but we should handle it.

I was thinking about using display configuration notification to notify that the display has become "off". But as I've discovered in a different occasion, a platform cannot force a display to go off, as a little thing called `DisplayConfigurationPolicy` will have a final say if a display will be on or off.

I have a feeling that, indeed, display configuration is where it should be handled ultimately. I'm just not sure how. `DisplayConfigurationPolicy` has a say on a number of settings there, so I'm not sure what I can use. Setting the output as "not connected" is probably not right.

Or maybe the DisplayConfiguration is just not the right place. I'm not even sure if Mir can stop compositing for a single output. This is all that I know from reading codes and discussions. I hope that help.

[1] For the avoidance of doubt, "platform" means Mir's server platform. I know that the distinction is not that important anymore given that client platform has gone with mirclient support, but just in case.

**Comment by AlanGriffiths** [comment]
I think the root cause of this is using the `DisplayClient::Output` implementation of `RenderTarget::swap_buffers()` call to wait for the frame notification.

AS @peat-psuwit says, we probably need more chatter between the render and the output to separate out the swapping of buffers and the indication that further rendering should commence.
